### PR TITLE
Add SPSAAttack

### PR DESCRIPTION
**Pull request by feinsteinben**
I've implemented the SPSA Attack in advertorch.

Please note that this implementation supports batch size >= 1, and not only batch size == 1.

The implementation includes addition of `MarginalLoss`, which was added to `advertorch.attacks.utils`. If there is a better place to put it in - please let me know.

The used optimizer is `Adam`. I've seen it being use in some other references, and it indeed achieved better results than `SGD`.

### CODE REVIEW

**Comment by tarokiritani** [comment]
@feinsteinben When `x` is expanded by a large `nb_sample` such as 8192 in [Uesato et al.](https://arxiv.org/pdf/1802.05666.pdf), the GPU memory can be easily depleted, or I would need tens of gpus.
There is no (algorithmic) reason to compute the forward pass all at once. Maybe we can input lets say 64 images many times, and later compute the mean? I suppose we can reorganize the code a bit and have a for loop or something in [`spsa_grad`](https://github.com/feinsteinben/advertorch/blob/spsa/advertorch/attacks/spsa.py#L41)?

**Comment by feinsteinben** [comment]
@tarokiritani I'll add an argument of `max_batch_size`, and a loop over batches of that limited size. Would that answer your concern?

**Comment by tarokiritani** [comment]
@feinsteinben If I can use your implementation with a small number of gpus, that would be great. Thanks!

**Comment by feinsteinben** [comment]
@tarokiritani I'm not sure how to support multi-gpus, but the current implementation should allow running this attack with large `nb_sample` (such as 8192) on a single gpu using small batches.

**Comment by gwding** [comment]
@feinsteinben sorry for the delayed response. I've made a pull request to your branch to add some testing and benchmarking.

Currently when I try to run the attack, it gives me the following error
```

**Comment by feinsteinben** [comment]
@gwding fixed this bug, and added SPSA to the tests.

**Comment by gwding** [comment]
@feinsteinben What do you think of the PRs on your branch? are you merging them? are having concerns? specifcally https://github.com/feinsteinben/advertorch/pull/1 and https://github.com/feinsteinben/advertorch/pull/2

**Comment by feinsteinben** [comment]
@gwding I wasn't aware of the two PRs. Thanks for letting me know about them.
* feinsteinben#2 was merged.
* feinsteinben#1 was independently solved, and can probably be solved. asked @tarokiritani to take a look at my fix.

File: advertorch/test_utils.py [hunk 1]
```diff
@@ -271,6 +276,7 @@ def generate_data_model_on_img():
     DDNL2Attack,
     SparseL1DescentAttack,
     L1PGDAttack,
+    # LinfSPSAAttack,  # XXX: not exactly sure: test says no
```
Comment (gwding) [inline advertorch/test_utils.py]: For me it seems that `LinfSPSAAttack` should be "batch consistent"? do you know which operation make it not consistent? or it is actually batch consistent, but the test fails because of some uncontrolled randomness?

Comment (feinsteinben) [reply advertorch/test_utils.py]: @gwding I also think it should be batch consistent. I'll look into it.

Comment (feinsteinben) [reply advertorch/test_utils.py]: @gwding I think it fails because the size of `v_` is not the same (so the randomness is different).

**Comment by feinsteinben** [comment]
@gwding maybe there is more to this than just different randomness. I think that the loss function should not be reduced in order for this attack to properly work on batches.

File: advertorch/test_utils.py (hunk 1 above)
Comment (tarokiritani) [reply advertorch/test_utils.py]: https://github.com/feinsteinben/advertorch/blob/bcbd733db448d7b5963e25153b73883a11fa4331/advertorch/attacks/spsa.py#L116-L119
v_ needs to be reinitialized here like x_ and y_. @feinsteinben I guess you are already aware of this, but just in case.

Comment (tarokiritani) [reply advertorch/test_utils.py]: https://github.com/feinsteinben/advertorch/blob/bcbd733db448d7b5963e25153b73883a11fa4331/advertorch/attacks/spsa.py#L116-L119
v_ needs to be reinitialized here like x_ and y_. I guess you are already aware of this, but just in case. Maybe related to this issue.

**Comment by gwding** [comment]
@feinsteinben are you talking about this line? https://github.com/BorealisAI/advertorch/pull/46/files#diff-c873222feb6394264bda2329a2dbd17bR157
I think `reduction='sum'` should be fine for batch, and it is implemented for other batch consistent attacks

**Comment by feinsteinben** [comment]
I think it's ok for *gradient-based* batch-consistent attacks, but not for *gradient-free* attacks.

**Comment by gwding** [comment]
@feinsteinben I guess that could be the case, what's your reasoning on this?
The reason that I can think of is that running on a single example vs a batch, even if they use the same random seed, the sequence that the random numbers are generated in a different way, so they won't be consistent. Do you think that's the reason? or is there something else?

**Comment by feinsteinben** [comment]
I think there are two, independent, problems:
1. The problem you stated. I tried to eliminate this problem (at least for tests) in my local branch, by generating the same noise for all the images in the batch and then expanding the batch dimension. However, some tests still didn't pass.
2. The reduction problem. In most _gradient-based_ attacks it's convenient to use `reduction="sum"` as the gradient w.r.t an image in the batch are not affected by the predictions on the other images in the batch (because there is no link in the network, so the gradient computed by back-propagation is zero). However, this is not the case when estimating the gradient numerically. Therefore, in this case, it's _crucial_ to compute the loss w.r.t to each image _independently_ (i.e. `reduction="none"`), otherwise the estimated gradient w.r.t an image _can_ be affected by the predictions of other images in the batch.

Problem #2 is not so easy to solve, as it requires to change the loss objects in the tests of the SPSA attack to have `reduction="none"`.

**Comment by gwding** [comment]
@feinsteinben I see. Thanks for the reasoning! I agree that `reduction="none"` is not something that we need to implement right now.

I think there're only few quick things before we can merge
1. could you try to run (or make it work) the file benchmark_spsa_attack.py , and paste the printout of performance report into the file? so we know what should be the performance of this implementation. An example is https://github.com/BorealisAI/advertorch/blob/master/advertorch_examples/attack_benchmarks/benchmark_iterative_projected_gradient.py
2. I'll add a PR to your branch to make sure your name appear in the copyright notices and authors' list (I forgot about this in your last PR, so will do it together)
3. This PR doesn't seem to be tested by travis-ci for some reason, and we might need to close this one and open a new one to make sure all tests pass.

**Comment by tarokiritani** [comment]
I think this line is wrong. I can fix it and send a PR hopefully today.
https://github.com/feinsteinben/advertorch/blob/52a6ad463f76b5b5f13fc6481d54c1f13464f797/advertorch/attacks/utils.py#L142
Here is a snippet using torch.index_select().
```
import torch
logits = torch.Tensor([[1, 2], [3, 4]])
targets = torch.Tensor([0, 1]).long()
print(torch.index_select(logits, -1, targets)) # tensor([1., 2.], [3., 4.])
```
I suppose you expect the output to be `tensor([1., 4.])`. @feinsteinben Correct me if I am mistaken.

**Comment by tarokiritani** [comment]
I looked into the batch consistency issue. This indeed seems to be caused by different randomization across batched inputs. By using the same random values in a batch, I made this algorithm batch consistent. @feinsteinben I sent you a [PR](https://github.com/feinsteinben/advertorch/pull/5).

**Comment by feinsteinben** [comment]
@tarokiritani I like your batch-consistency fix. I suggested a way to reduce the memory consumption and run time, and would appreciate if you could take a look at my suggestions. Thanks for the help!

**Comment by feinsteinben** [comment]
thanks to @tarokiritani, the attack is now batch-consistent.

**Comment by feinsteinben** [comment]
@gwding I think this PR is ready for merge. I think it's time for a final review.
@tarokiritani thanks for your tremendous help!

**Comment by gwding** [comment]
@feinsteinben @tarokiritani Thanks for the hard work!
I made a PR https://github.com/feinsteinben/advertorch/pull/9 adding some final changes to this branch before it can be merged to master.

**Comment by feinsteinben** [comment]
@gwding "allow edits from maintainers" is now enabled. You can now edit whatever is needed for merge.

**Comment by feinsteinben** [comment]
thx! :-)

**Comment by gwding** [comment]
@feinsteinben You're welcome and thanks again for tons of work and efforts :)
